### ISSUE TITLE
Teams desktop and iOS inconsistent image rendering behavior

### ISSUE THREAD

**Issue by stpeteparadox** [MicrosoftDocs/msteams-docs#388]
I am testing a bot proactively sending status updates / notifications to a Teams channel (not using incoming webhook as it doesn't support adaptive cards). The notification is in the form of an adaptive card with a title, image, and message:

```
card.Body.Add(new AdaptiveTextBlock()
{
  Text = "A title here",
  Size = AdaptiveTextSize.Large,
  Weight = AdaptiveTextWeight.Bolder,
  Wrap = true
});
card.Body.Add(new AdaptiveImage()
{
  Type = "Image",
  Size = AdaptiveImageSize.Auto,
  Style = AdaptiveImageStyle.Default,
  HorizontalAlignment = AdaptiveHorizontalAlignment.Center,
  Separator = true,
  Url = "https://via.placeholder.com/1000"
});
card.Body.Add(new AdaptiveTextBlock()
{
  Text = "A message here",
  Wrap = true
});
```

When viewing this card in Teams on iOS, the image is clickable, expanding to fill the entire screen. However, when viewing the same card via Teams on Windows desktop, the image doesn't expand and is not clickable. Ideally, the image on Teams desktop would expand to its larger, original size.

If I instead send the image as a pure message attachment, we see the opposite behavior.

```
Attachment attachment = new Attachment()
{
  ContentType = "image/png",
  ContentUrl = "https://via.placeholder.com/1000"
};
```

In this case, the image on Teams desktop renders slightly smaller, but is clickable, expanding to a larger size, and then to its largest, original size. This is the ideal behavior. However, on iOS, when tapping the image, it does not expand, and instead the message loads in a separate window, showing the image in the same size in the message body.

We are attempting to convert over to Teams from Slack but this behavior is likely to hold us back as the adaptive card image will be unreadable in desktop format. Thank you

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 40f35d71-c372-14ab-17d3-f4fc2a241a62
* Version Independent ID: f94dbfa0-9e4f-31cc-26ab-48dc3dda993a
* Content: [Text formatting in cards - Teams](https://docs.microsoft.com/en-us/microsoftteams/platform/concepts/cards/cards-format)
* Content Source: [msteams-platform/concepts/Cards/cards-format.md](https://github.com/MicrosoftDocs/msteams-docs/blob/master/msteams-platform/concepts/Cards/cards-format.md)
* Product: **msteams**
* GitHub Login: @o365devx
* Microsoft Alias: **o365devx**

**Comment by Wajeed-msft** [comment]
Could you please share both the screenshots?

**Comment by stpeteparadox** [comment]
Hi @Wajeed-msft , please see screenshots below, for both AdaptiveCard/Image approach (preferred) and Attachment approach, both via a proactive bot.

**_AdaptiveCard approach_**
_Teams on Windows_
![windows-teams-adaptivecard](https://user-images.githubusercontent.com/3724281/52723078-d0e30300-2fe7-11e9-8248-f0cabfc98633.png)
In the above, the AdaptiveImage _is not_ expandable. This is the primary problem we are facing. Ideally this should expand in the same manner that an image sent as an Attachment does (and offering further zoom tools in top-right).

_Teams on iOS_
![ios-teams-adaptivecard](https://user-images.githubusercontent.com/3724281/52723260-25867e00-2fe8-11e9-842d-f0ef870ef61b.png)

![ios-teams-adaptivecard-expandedimagefromcard](https://user-images.githubusercontent.com/3724281/52723275-2c14f580-2fe8-11e9-8caf-79d56ce5f5da.png)
AdaptiveCard and expandable AdaptiveImage seem to work fine on Teams on iOS.


**_Attachment approach_**
_Teams on Windows_
![windows-teams-attachment](https://user-images.githubusercontent.com/3724281/52723390-61b9de80-2fe8-11e9-9f44-5844990f8b9c.png)

![windows-teams-attachment-expandedfromattachment](https://user-images.githubusercontent.com/3724281/52723401-68e0ec80-2fe8-11e9-9ed9-0d43470698a5.png)
Expandability of image attachment on Teams on Windows works, but unfortunately this approach doesn't have the formatting capabilities that are found in with an AdaptiveCard.

_Teams on iOS_
![ios-teams-attachment](https://user-images.githubusercontent.com/3724281/52723503-99c12180-2fe8-11e9-9c47-5d2670bbfa08.png)

![ios-teams-attachment-onfirsttap](https://user-images.githubusercontent.com/3724281/52723512-a04f9900-2fe8-11e9-9c25-2502d484086d.png)

![ios-teams-attachment-expandedonsecondtap](https://user-images.githubusercontent.com/3724281/52723529-a776a700-2fe8-11e9-9d0d-c90b7bc977c0.png)
Expandability of image attachment on Teams on iOS also technically works, but requires the user to tap twice. First to get to an intermediate page (see second image for iOS attachment approach), and then a second time to fill the image to the screen.


_**Summary**_
We would love to be able to simply use the AdaptiveCard approach and have consistent image expandability behavior across Teams for Windows Desktop, iOS and Android (though we haven't tested on the latter). At the moment, AdaptiveImage expandability is not working in Teams on Windows in the manner that an image sent as an attachment would.

This is the final item keeping us in Slack.

**Comment by stpeteparadox** [comment]
Hi Microsoft,

Any update on a fix for this inconsistent image rendering behavior? We continue to see it occurring in our test Teams accounts. 

Many thanks for your help :)

**Comment by Wajeed-msft** [comment]
Thanks for reporting this. I will get back to you on this.

**Comment by stpeteparadox** [comment]
Hi Microsoft / @Wajeed-msft : 

Please let us know if there is anything we can further do to help on this one. The issue remains: In Teams on Windows, when clicking an image sent as an AdaptiveImage within the body of an AdaptiveCard, it does not expand to a larger size. However, when clicking an image sent purely as an Attachment (not as part of an AdaptiveCard), it is clickable/expandable. 

In both cases the exact same image is sent. The hover-over mouse cursor is different in each case, with the former just being the normal Windows mouse icon, and in the latter case it being a human hand with index finger extended.

It's really important that our users are able to expand and read the image sent as part of an AdaptiveCard. We cannot use the pure Attachment approach as this method lacks the rich text formatting features available in an AdaptiveCard.

Thank you

**Comment by stpeteparadox** [comment]
Here is the json for the AdapativeCard, in case of convenience:

```
{
    "type": "AdaptiveCard",
    "body": [
        {
            "type": "TextBlock",
            "size": "Large",
            "weight": "Bolder",
            "text": "A title here",
            "wrap": true
        },
        {
            "type": "Image",
            "horizontalAlignment": "Center",
            "separator": true,
            "url": "https://via.placeholder.com/1000"
        },
        {
            "type": "TextBlock",
            "text": "A message here: \n 1)  Sub-point 1 \n 2)  Sub-point  2 \n 3)  Sub-point 3 \n 4)  Sub-point 4 \n 5)  Sub-point 5",
            "wrap": true
        }
    ],
    "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
    "version": "1.0"
}
```

**Comment by stpeteparadox** [comment]
Hi Microsoft / @Wajeed-msft ,

Any update on this bug / issue? 

Most appreciated! Thanks

**Comment by siduppal** [comment]
@stpeteparadox can you use an OpenUrl select-action on the image?

```json
{
    "type": "AdaptiveCard",
    "body": [
        {
            "type": "TextBlock",
            "size": "Large",
            "weight": "Bolder",
            "text": "A title here",
            "wrap": true
        },
        {
            "type": "Image",
            "horizontalAlignment": "Center",
            "separator": true,
            "url": "https://via.placeholder.com/1000",
            "selectAction": {
                                "type": "Action.OpenUrl",
                                "url": "https://via.placeholder.com/1000"
                            }
        },
        {
            "type": "TextBlock",
            "text": "A message here: \n 1)  Sub-point 1 \n 2)  Sub-point  2 \n 3)  Sub-point 3 \n 4)  Sub-point 4 \n 5)  Sub-point 5",
            "wrap": true
        }
    ],
    "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
    "version": "1.0"
}
```

**Comment by stpeteparadox** [comment]
Hi Microsoft,

My team and I didn't go to the effort of testing and documentation for this post out of only pure love for Microsoft and technology. We have very real business needs. We are very real _paying_ customers of your M365 services.

Unfortunately it continues to appear that no one is taking our issue very seriously, or at least not appreciating what exactly is the problem. Perhaps it hasn't yet been replicated. Even a decision / reply saying that "we have replicated the issue you describe, but we have decided that we are actually ok with the inconsistent user experience for these {...} reasons" would be sufficient for us to move on/forward--as sad as that nevertheless would be.

@siduppal Thank you. We do appreciate your suggestion. However, the workaround you suggest is not a fix and isn't new, but rather is an entirely different way of using/accessing an image sent within an AdaptiveCard in Teams. When you apply a select-action, it forces the linked-to item to be opened in a separate browser. If doing so on a Windows desktop (using even the new Edge browser), the file doesn't get opened in the browser and instead is downloaded as a file. And even worse, in iOS where AdaptiveImage built-in expandability is actually working today, the image instead is opened separately in a browser window. To put it kindly, this is a sub-optimal user experience.

I understand that Microsoft is up to 1 million and 1 different things and perhaps many of you wear more than one hat across even multiple product teams, whereas someone like Slack has a much more focused approach (more so a singular product) allowing them to monitor and improve the small, but critical user experience details. But this is a problem ... Microsoft has long had a reputation for big picture, while light on detail/implementation; we all really do hope that can change/improve with your greater focus on a more service-based, platform approach. 

Please let us describe again how things ought to work when an image (AdaptiveImage) is sent by a pro-active bot in an AdaptiveCard, and viewed via Teams across devices (e.g. Windows and iOS):

- Windows: if the sent image is bigger than the AdaptiveCard size (e.g. in the above case a 1k x 1k px image), then it should be expandable when clicked on. Note that this is not an unheard-of feature. Currently when, either as a coded bot or as a human-user, one attaches an image to a message, that image will then be expandable when clicked on in Teams for Windows (there is even a further zoom feature available). It doesn't open in a browser window, it opens as an expanded overlay window within Teams-just like how it has almost always worked in Slack. Currently, the AdaptiveImage in an AdaptiveCard in Teams for Windows is not expandable at all. So this is inconsistent. Please read on for the second inconsistency.

- iOS: if that same image sent as per the Windows scenario is instead viewed via Teams for iOS, the AdaptiveImage is indeed expandable, just like the attachment version of it is. It expands to fill the screen and can be further zoomed. So here, the AdaptiveImage is expandable on iOS, but it is not expandable on Windows...

I know Teams has no intentions to match Slack 1-for-1 and you have your own vision, but this is not something special we are asking. Rather **_we are asking for something more important than constantly new features, we are asking for a consistent design/user experience._** There's got to be a balance always, but please Microsoft slow down a bit and consolidate your feature-creep and bring basic-expected quality and consistency back up to respectable levels, then start again on new features.

Thank you for your help on resolving this matter.

**Comment by siduppal** [comment]
Thank you @stpeteparadox for your detailed response explaining the issue. I understand the importance of this issue and the pain caused by inconsistencies in rendering of Adaptive Cards across different Teams clients. Reading your response, I am filled with a feeling of gratitude to have passionate developers like you to serve. 

The pop-over behavior for images wasn't part of the AdaptiveImage spec and that's where the inconsistency is originating from. I will follow up with our team on enhancing the spec to handle this scenario. In the meanwhile, you could consider using the OpenUrl workaround I suggested, though I understand that's not ideal for your scenario and not exactly equivalent.

I would invite you to review and comment on the inconsistencies spec to address all such issues in Adaptive Cards.

* [Inconsistencies spec](https://github.com/microsoft/AdaptiveCards/issues/3583)
* [iOS / Android inconsistencies spec](https://github.com/microsoft/AdaptiveCards/issues/3653)

Please do not hesitate to reach out with more issues in the future! I see that we didn't really do a good job on following up on this issue originally logged a while back, but that's not the norm we aspire for. You may also utilize other developer support options listed [here](https://developer.microsoft.com/en-us/microsoft-teams/docs#microsoftteamscommunity).

**Comment by siduppal** [comment]
Hi @stpeteparadox - I hope you are staying healthy in these unprecedented times. I just wanted to drop you an update to share that we are addressing the item you brought up on Teams web/desktop for images in Adaptive Cards. Users will be able to hover over an image in Adaptive Card and tap on an icon to expand it. 

![captured (3)](https://user-images.githubusercontent.com/7799064/78404525-7aea6980-75b3-11ea-9d23-a64c95881b23.gif)

I also wanted to share that the delay in responding to your bug made us reevaluate the process and setup automated monitoring of items here using MSFTBot. We will continue to tweak the process here to ensure that the issues are addressed quickly in a satisfactory fashion.

Again, I hope you stay healthy through these times and after. Please do not hesitate to reach out with any questions/concerns via this GitHub repo or any of the other support options listed [here](https://developer.microsoft.com/en-us/microsoft-teams/docs#microsoftteamscommunity), or messaging me.

Thanks!

**Comment by stpeteparadox** [comment]
Hi @siduppal 

Thank you and likewise wishing you, your family and Microsoft colleagues much health and best wishes.

This is wonderful news. Thank you very much. Any thoughts on when this might be available even if in preview form?

Prior to your reply, our team went ahead and converted the underlying content that is normally transmitted in the adaptive-card via a screenshot, into a full adaptive-card json format, with interactive expanding sections. However, we discovered that while this card works fine in the desktop version of Teams, in Teams for iOS (iPhone 7) it loads in a very, very sluggish manner--not at all fluid. We are probably exceeding whatever the current best-practice guidelines are for size and breadth of content in json format, so your latest update about the fix for image expandability is really, really good news.

Of course, if you want to know more about our rather sluggish adaptive-card json version of the content normally being sent as an image, we are more than happy to share after some censorship.

Thank you again and stay safe!

**Comment by siduppal** [comment]
@stpeteparadox we are hoping to get this out in GA around Build. Pls feel free to start a separate GitHub issue thread on the sluggish problem you noticed and we will investigate that one too.

Cheers!

**Comment by siduppal** [comment]
@stpeteparadox we received some strong feedback from some internal dogfood users about making all images in Adaptive Cards expandable by default so we are now changing it be opt-in on developer side (you'd simply set a property on image to indicate you want it to be expandable). The new ETA is August.

**Comment by stpeteparadox** [comment]
Hi @siduppal thank you for this encouraging and exciting update. Would it be reasonable to assume that this 'expandable image' toggle property will also have effect if the adaptive card was received via an incoming webhook (#3740 - still a message attachment) in Teams, as opposed to having been sent via a bot application?

**Comment by siduppal** [comment]
@stpeteparadox this is now available in developer preview on web and desktop. To make an image expandable, enable the `allowExpand` flag as shown below.

```json
    {
      "type": "Image",
      "url": "https://picsum.photos/200/200?image=110",
      "msTeams": {
        "allowExpand": true
      }
    }
```
This will cause the client to render an element on hover to allow the user to expand the image.

![image](https://user-images.githubusercontent.com/7799064/90681872-f5d9b700-e218-11ea-8ef7-bcf88c49a39c.png)

This behavior is supported for Incoming Webhooks as well.

Please try it out and share your feedback!

**Comment by stpeteparadox** [comment]
Thank you, thank you, thank you! I am so happy we now have this small, but rather useful/practical feature in Teams. Very much appreciated @siduppal for your help with this and for providing updates over time. Thanks again.

Fingers now crossed for first-class citizen treatment of private channels (ie support for bots, etc).

**Comment by stpeteparadox** [comment]
Hi @siduppal, 

We'd like to update you on one odd behavior we've noticed: the hover-over expand button goes away when we switch out of the channel where the image has been posted, to another section of Teams (ie Chat, etc) and then come back to it. Or if we switch to another application, then come back. The button will not re-appear. 

The only way we've been able to get it back is to switch to a separate channel with the Teams section (can be a channel in the same Team or a different Team), then the button will re-appear.

FYI
